game-core/src/test/java/games/strategy/triplea/delegate/battle/steps/change/MarkNoMovementLeftTest.java

```diff
@@ -22,24 +35,84 @@
   @Mock BattleActions battleActions;
 
   @Test
-  void runsOnFirstRound() {
-    final BattleState battleState = givenBattleStateBuilder().battleRound(1).build();
+  void doesNotRunOnFirstRoundAndHeadless() {
+    final BattleState battleState = givenBattleStateBuilder().battleRound(1).headless(true).build();
     final MarkNoMovementLeft markNoMovementLeft =
         new MarkNoMovementLeft(battleState, battleActions);
 
     markNoMovementLeft.execute(executionStack, delegateBridge);
 
-    verify(battleActions).markNoMovementLeft(eq(delegateBridge));
+    verify(delegateBridge, never()).addChange(any());
   }
 
   @Test
   void doesNotRunOnSecondRound() {
-    final BattleState battleState = givenBattleStateBuilder().battleRound(2).build();
+    final BattleState battleState =
+        givenBattleStateBuilder().battleRound(2).headless(false).build();
     final MarkNoMovementLeft markNoMovementLeft =
         new MarkNoMovementLeft(battleState, battleActions);
 
     markNoMovementLeft.execute(executionStack, delegateBridge);
 
-    verify(battleActions, never()).markNoMovementLeft(eq(delegateBridge));
+    verify(delegateBridge, never()).addChange(any());
+  }
+
+  @Test
+  void nonAirWithMovementLeftAreMarkedAsMoved() {
+    final BattleState battleState =
+        givenBattleStateBuilder()
+            .battleRound(1)
+            .headless(false)
+            .attackingUnits(List.of(givenNonAirUnitWithMovementLeft(BigDecimal.ONE)))
+            .build();
+    final MarkNoMovementLeft markNoMovementLeft =
+        new MarkNoMovementLeft(battleState, battleActions);
+
+    markNoMovementLeft.execute(executionStack, delegateBridge);
+
+    verify(delegateBridge).addChange(any());
```

Comment:
The change object is the interesting thing, right? The structure and data of the change object we do expect to be in a certain way, that defines correctness? For example, if it were null, I think this test might still pass.

A captor could be used to grab the change object and then verify it. We are then pretty tied to implementation and we have another code smell where the code could change, still be correct, and the tests are breaking because they were testing implementation and not behavior.

So, captor is an option, another option is to have the `execute` method return the change object. In that case the test code would get a reference to it directly. IF we are limited by compatibility concerns (we might not be), but if we are, then the existing code could be converted to a facade and invokes a method that is visible to testing that does return a change object. 

WDYT?

Reply:
I changed the code to use `any(Change.class)`.  That prevents null.

As for actually verifying the change object, I think that might get really tied to the implementation.  This particular change object is a `CompositeChange` with multiple `ObjectPropertyChange` changes in it.  So the test would have to check that it is a `CompositeChange` and then grab the inner changes, cast them to `ObjectPropertyChange`, and then verify that they are setting the correct property (`ALREADY_MOVED`) and that the new value and the old value (which currently have no getters) are what is expected.  Is there already a change matcher somewhere in the code base that would do this?

Another idea is to just compare the `toString` result of the change object.  Then the code would be something like:
```
    verify(delegateBridge).addChange(argThat(hasToString("CompositeChange <[Property change, unit:null owned by null property:alreadyMoved newValue:1 oldValue:0]>")));
```

Reply:
Indeed, it sounds like the check is if the expected property and perhaps expected value are present anywhere in the change object. A matcher would be pretty slick since you could do an iterative search through the change object. 'toString' is a bit brittle as you rely on the the toString method, but it would work too. IMO the matcher is probably the more robust way to go. AFAIK there is no such matcher.

Reply:
I've made a custom matcher for `ObjectPropertyChange` and `CompositeChange`.  As more `Change` objects are used, the matcher can be expanded.